packager/mux/mux.go

```diff
@@ -37,82 +37,134 @@ import (
+// particular prefix should precede the rule for it's sub-prefix. E.g. the rule
+// for routing requests prefixed with “/priv/doc/” (w/ trailing slash) must go
+// before the rule for routing requests prefixed with “/priv/doc” (w/o trailing
+// slash).
+// The last rule in the matrix must have empty URLPathPrefix, so it would match
+// any URL. This ensures there's at least one matching rule for any URL.
 type mux struct {
-	certCache   http.Handler
-	signer      http.Handler
-	validityMap http.Handler
-	healthz     http.Handler
+	routingMatrix []routingRule
+}
+
+// return404 is a URL Path Suffix Validator that always returns 404.
+func return404(suffix string, req *http.Request, params *map[string]string, errorMsg *string, errorCode *int) {
+	*errorMsg, *errorCode = "404 page not found", http.StatusNotFound
+}
+
+// expectNoSuffix is a URL Path Suffix Validator that expects an empty suffix.
+func expectNoSuffix(suffix string, req *http.Request, params *map[string]string, errorMsg *string, errorCode *int) {
+	if suffix != "" {
+		return404(suffix, req, params, errorMsg, errorCode)
+	}
+}
+
+// expectSignerQuery is a URL Path Suffix Validator specific to signer requests.
+func expectSignerQuery(suffix string, req *http.Request, params *map[string]string, errorMsg *string, errorCode *int) {
+	(*params)["signURL"] = suffix
+	if req.URL.RawQuery != "" {
+		(*params)["signURL"] += "?" + req.URL.RawQuery
+	}
+}
+
+// expectCertQuery is a URL Path Suffix Validator specific to cert requests.
+func expectCertQuery(suffix string, req *http.Request, params *map[string]string, errorMsg *string, errorCode *int) {
+	unescaped, err := url.PathUnescape(suffix)
+	if err != nil {
+		*errorMsg, *errorCode = "400 bad request - bad URL encoding", http.StatusBadRequest
+	} else {
+		(*params)["certName"] = unescaped
+	}
 }
 
-// The main entry point. Use the return value for http.Server.Handler.
+// New is the main entry point. Use the return value for http.Server.Handler.
 func New(certCache http.Handler, signer http.Handler, validityMap http.Handler, healthz http.Handler) http.Handler {
-	return &mux{certCache, signer, validityMap, healthz}
+	return &mux{
+		// Note that the order of rules in the matrix matters: the first
+		// matching rule will be applied, so the rule for “/priv/doc/” precedes
+		// the rule for “/priv/doc”.
+		// Also note that the last rule matches any URL.
+		[]routingRule{
+			{"/priv/doc/", expectSignerQuery, signer, "signer"},
+			{"/priv/doc", expectNoSuffix, signer, "signer"},
+			{"/amppkg/cert/", expectCertQuery, certCache, "certCache"},
+			{"/amppkg/validity", expectNoSuffix, validityMap, "validityMap"},
+			{"/healthz", expectNoSuffix, healthz, "healthz"},
+			{"", return404, nil, "handler_not_assigned"},
+		}}
 }
 
 func tryTrimPrefix(s, prefix string) (string, bool) {
 	sLen := len(s)
 	trimmed := strings.TrimPrefix(s, prefix)
-	return trimmed, len(trimmed) != sLen
+	return trimmed, len(prefix)+len(trimmed) == sLen
```

Comment:
Do you have an example (or test) to demonstrate the need for this change? I'm not seeing it; AFAICT [TrimPrefix](https://golang.org/pkg/strings/#TrimPrefix) only returns the original string or the correctly-trimmed string.

Also, the previous version seemed clearer to me, but I could be convinced otherwise.